src/std/make.ss

```diff
@@ -20,224 +41,388 @@
-    (let* ((mod (import-module file))
-           (ht  (make-hash-table-eq))
-           (pre (core-context-prelude mod)))
-      (let lp ((rest (cons pre (module-context-import mod))))
-        (match rest
-          ([hd . rest]
-           (cond
-            ((module-context? hd)
-             (hash-put! ht (expander-context-id hd) #t)
-             (lp rest))
-            ((prelude-context? hd)
-             (alet (id (expander-context-id hd)) ; maybe it's root!
-               (hash-put! ht id #t))
-             (lp rest))
-            ((module-import? hd)
-             (lp (cons (module-import-source hd) rest)))
-            ((module-export? hd)
-             (lp (cons (module-export-context hd) rest)))
-            ((import-set? hd)
-             (lp (cons (import-set-source hd) rest)))
-            (else
-             (error "Unexpected module import" hd))))
-          (else
-           [file (expander-context-id mod) (sort (hash-keys ht) symbol<?) ...])))))
-  (map depgraph files))
-
-;; make-depgraph/spec : buildspec -> depgraph
-(def (make-depgraph/spec spec)
-  (make-depgraph (buildspec-depfiles spec)))
-
-;; buildspec-depfiles : buildspec -> (listof file)
-;; Produces the list of files with deps relevant for making a depgraph
-(def (buildspec-depfiles spec)
-  (def (file-e mod ext)
-    (if (string-empty? (path-extension mod))
-      (string-append mod ext)
-      mod))
-
-  (let lp ((rest spec) (files []))
-    (match rest
-      ([hd . rest]
-       (match hd
-         ((? string?)
-          (lp rest (cons (file-e hd ".ss") files)))
-         ([gxc: mod . opts]
-          (lp rest (cons (file-e mod ".ss") files)))
-         ([exe: mod . opts]
-          (lp rest (cons (file-e mod ".ss") files)))
-         ([static-exe: mod . opts]
-          (lp rest (cons (file-e mod ".ss") files)))
-         ([ssi: mod . opts]
-          (lp rest (cons (file-e mod ".ssi") files)))
-         (else
-          ; otherwise it's static-include: or copy:, no deps,
-          ; not relevant for making a depgraph
-          (lp rest files))))
-      (else
-       (reverse files)))))
+;;; Functions that should be better moved some library...
+(defrule (when/list cond list) (if cond list []))
+(def (listify x) (when/list (pair? x) x))
+
+(def (prefix/ prefix path)
+  (if prefix
+    (string-append prefix "/" path)
+    path))
+
+(def (force-outputs) (force-output (current-error-port)) (force-output)) ;; misc/ports
+(def (message . lst) (apply displayln lst) (force-outputs))
+(def (writeln x) (write x) (newline) (force-outputs)) ;;
+
+;;; Functions partially reimplemented from std/srfi/43
+(def (vector-for-each f v)
+  (def l (vector-length v))
+  (let loop ((i 0)) (when (< i l) (begin (f i (vector-ref v i)) (loop (+ 1 i))))))
+(def (vector-ensure-ref v i f)
+  (or (vector-ref v i) (let ((x (f))) (vector-set! v i x) x)))
+(def (vector-map f v)
+  (def w (make-vector (vector-length v)))
+  (vector-for-each (lambda (i x) (vector-set! w i (f x))) v)
+  w)
```

Comment:
let's add a TODO here with a bug pointer, so that we can eventually get rid of these defs.

Reply:
Done.